Enhanced <input type=color>: make a11y code account for alpha

annevk opened the pull request:
#### 4cb42b8c9386d435841b39b4864d920ee9416c03
<pre>
Enhanced &lt;input type=color&gt;: make a11y code account for alpha
<a href="https://bugs.webkit.org/show_bug.cgi?id=281283">https://bugs.webkit.org/show_bug.cgi?id=281283</a>
<a href="https://rdar.apple.com/137734496">rdar://137734496</a>

Reviewed by Tyler Wilcock.

Change the fallback color from transparent black to opaque black to
preserve how it is exposed today in accessibility. This also happens to
match the fallback color used by the control itself. It is not entirely
clear why transparent black was used before.

And start serializing the alpha component of the color in
AXCoreObject::value() when it is not 1. As all existing code paths can
only hit it when it is 1 there is no preference guard here.

* LayoutTests/accessibility/color-well-expected.txt:
* LayoutTests/accessibility/color-well-legacy-expected.txt: Copied from LayoutTests/accessibility/color-well-expected.txt.
* LayoutTests/accessibility/color-well-legacy.html: Copied from LayoutTests/accessibility/color-well.html.
* LayoutTests/accessibility/color-well.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/accessibility/color-well-legacy-expected.txt: Renamed from LayoutTests/platform/gtk/accessibility/color-well-expected.txt.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/accessibility/color-well-legacy-expected.txt: Renamed from LayoutTests/platform/wpe/accessibility/color-well-expected.txt.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::value):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::colorValue const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::colorValue const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::valueAsColor const):

Canonical link: <a href="https://commits.webkit.org/285395@main">https://commits.webkit.org/285395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b287aae6c6262b2fa6123b885c4ba855becab37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72429 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25217 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76610 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23638 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74544 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59654 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23460 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57036 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15551 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46944 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62371 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37472 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43595 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19838 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21988 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65477 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78277 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16670 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19335 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65489 "Found 4 new test failures: fast/editing/document-leak-altered-text-field.html webgl/2.0.y/conformance2/textures/webgl_canvas/tex-3d-r11f_g11f_b10f-rgb-float.html webgl/2.0.y/conformance2/textures/webgl_canvas/tex-3d-r11f_g11f_b10f-rgb-unsigned_int_10f_11f_11f_rev.html webgl/2.0.y/conformance2/textures/webgl_canvas/tex-3d-rgba16f-rgba-float.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16718 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64755 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13028 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6663 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11131 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47648 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2433 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48717 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48460 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/color-well-legacy.html:
```diff
@@ -0,0 +1,29 @@
+<input id="good_colorwell" type="color" value="#ff0000">
+<input id="bad_colorwell" type="color" value="purple">
+
+<script>
+    description("This test checks the role of ColorWellRole on an input with type=color");
```
In newer accessibility tests, we aren't indenting a level inside a `<script>` tag, so everything inside can move back one tab.

twilco commented on LayoutTests/accessibility/color-well-legacy.html:
```diff
@@ -0,0 +1,29 @@
+
+    if (window.accessibilityController) {
+        let axColorwell = accessibilityController.accessibleElementById("empty_colorwell");
+        debug("Role of input type=color is: " + axColorwell.role);
```
In newer accessibility tests, we don't use `debug` or `description`, as they write to the DOM immediately, which has made a noticeable difference in test speed (though less so in smaller tests like this). Instead, we would declare:

```
var output = "This test checks the role of ColorWellRole on an input with type=color\n\n";
```

Where `description` is, and replace `debug` with `output += "foo\n";`

And then `debug(output)` at the end.

Here's an example of a test following this format:

https://github.com/WebKit/WebKit/blob/main/LayoutTests/accessibility/changing-aria-hidden-with-display-none-parent.html#L16

twilco commented on LayoutTests/accessibility/color-well-legacy.html:
```diff
@@ -0,0 +1,29 @@
+        debug("Value of empty color well: " + axColorwell.stringValue);
+
+        axColorwell = accessibilityController.accessibleElementById("good_colorwell");
+        debug("Value of good color well: " + axColorwell.stringValue);
```
Take or leave this, but we could also consider using JS template strings vs. `+` style concatenation:

```
output += `Value of good color well: ${axColorwell.stringValue}\n`;
```

twilco commented on LayoutTests/accessibility/color-well.html:
```diff
@@ -23,8 +25,13 @@
         debug("Value of bad color well: " + axColorwell.stringValue);
+
+        axColorwell = accessibilityController.accessibleElementById("alpha_colorwell");
+        debug("Value of alpha color well: " + axColorwell.stringValue);
```
Fine to leave these `debug`s as-is since this is an existing test.

annevk committed:
Enhanced <input type=color>: make a11y code account for alpha
https://bugs.webkit.org/show_bug.cgi?id=281283
rdar://137734496

Reviewed by Tyler Wilcock.

Change the fallback color from transparent black to opaque black to
preserve how it is exposed today in accessibility. This also happens to
match the fallback color used by the control itself. It is not entirely
clear why transparent black was used before.

And start serializing the alpha component of the color in
AXCoreObject::value() when it is not 1. As all existing code paths can
only hit it when it is 1 there is no preference guard here.

* LayoutTests/accessibility/color-well-expected.txt:
* LayoutTests/accessibility/color-well-legacy-expected.txt: Copied from LayoutTests/accessibility/color-well-expected.txt.
* LayoutTests/accessibility/color-well-legacy.html: Copied from LayoutTests/accessibility/color-well.html.
* LayoutTests/accessibility/color-well.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/accessibility/color-well-legacy-expected.txt: Renamed from LayoutTests/platform/gtk/accessibility/color-well-expected.txt.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/accessibility/color-well-legacy-expected.txt: Renamed from LayoutTests/platform/wpe/accessibility/color-well-expected.txt.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::value):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::colorValue const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::colorValue const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::valueAsColor const):

Canonical link: https://commits.webkit.org/285395@main
